HARK/dcegm.py

```diff
@@ -8,428 +8,298 @@
-        see description above
-    """
-    # NOTE: assumes that the first segment is in fact increasing (forced in EGM
-    # by augmentation with the constrained segment).
-    # elements in common grid g
-
-    # Identify index intervals of falling and rising regions
-    # We need these to construct the upper envelope because we need to discard
-    # solutions from the inverted Euler equations that do not represent optimal
-    # choices (the FOCs are only necessary in these models).
-    #
-    # `fall` is a vector of indeces that represent the first elements in all
-    # of the falling segments (the curve can potentially fold several times)
-    fall = np.empty(
-        0, dtype=int
-    )  # initialize with empty and then add the last point below while-loop
-
-    rise = np.array([0])  # Initialize such thatthe lowest point is the first grid point
-    i = 1  # Initialize
-    while i <= len(x) - 2:
-        # Check if the next (`ip1` stands for i plus 1) grid point is below the
-        # current one, such that the line is folding back.
-        ip1_falls = x[i + 1] < x[i]  # true if grid decreases on index increment
-        i_rose = x[i] > x[i - 1]  # true if grid decreases on index decrement
-        val_fell = v[i] < v[i - 1]  # true if value rises on index decrement
-
-        if (ip1_falls and i_rose) or (val_fell and i_rose):
-
-            # we are in a region where the endogenous grid is decreasing or
-            # the value function rises by stepping back in the grid.
-            fall = np.append(fall, i)  # add the index to the vector
-
-            # We now iterate from the current index onwards until we find point
-            # where resources rises again. Unfortunately, we need to check
-            # each points, as there can be multiple spells of falling endogenous
-            # grids, so we cannot use bisection or some other fast algorithm.
-            k = i
-            while x[k + 1] < x[k]:
-                k = k + 1
-            # k now holds either the next index the starts a new rising
-            # region, or it holds the length of M, `m_len`.
-
-            rise = np.append(rise, k)
-
-            # Set the index to the point where resources again is rising
-            i = k
+            ends.append(i - 1)
+            starts.append(i)
 
         i = i + 1
 
-    # Add the last index for convenience (then all segments are complete, as
-    # len(fall) == len(rise), and we can form them by range(rise[j], fall[j]+1).
-    fall = np.append(fall, len(v) - 1)
+    # The last segment always ends in the last point
+    ends.append(len(y) - 1)
 
-    return rise, fall
+    starts = np.array(starts)
+    ends = np.array(ends)
 
+    return starts, ends
 
-# think! nanargmax makes everythign super ugly because numpy changed the wraning
-# in all nan slices to a valueerror...it's nans, aaarghgghg
 
-
-def calc_linear_crossing(m, left_v, right_v):
+def upper_envelope(segments, calc_crossings=True):
```

Comment:
Could this be numba-fied as well? are there drawbacks?

Reply:
No real drawbacks that I can think of. Just a lot of frustration about figuring out the exact types of inputs and outputs and how to make numba understand them.
"segments" for instance is a list of tuples of np arrays of floats. I need to figure out how to do that type declaration.